Feature/ Handle V3 sample specification

annakrystalli opened the pull request:
This PR implement and adds new tests for checking the validity of submissions of samples using the v3 schema sample spec. See [v3 sample validation spec for details](https://docs.google.com/document/d/1vJT8Uw8fQgoY5sajcokZSfnF7M89eiiiwlRXVkCuj2k/edit#heading=h.qli9sktijlj4).

## Specific Sample validation tests implemented (#80)

- [x] Validate all value combinations valid (as part of `check_tbl_values()`)
- [x] Validate all required value combinations submitted (as part of `check_tbl_values_required()`)
- [x] Validate correct number of samples per compound idx submitted. Added through function `check_tbl_spl_n()`.
- [x] Validate that samples within a submission file contain the same combination of optional (non-compound task id) values across all samples. Added through function `check_tbl_spl_non_compound_tid()`.
- [x] Validate that samples conform to sample dependence defined by the compound task id set configuration i.e. all samples for a given compound idx contain the same unique combination of compound task id values. Added through function `check_tbl_spl_compound_tid()`.
- [x] Add new validation checks to validate_model_data() in a back compatible way (i.e. only deploy if using a v3 config).

The key to the new functions `check_tbl_spl_n()`, `check_tbl_spl_non_compound_tid()` and `check_tbl_spl_compound_tid()` is a table of hashes on model output data joined to the output of the new `hubData::expand_model_out_val_grid(include_sample_ids = TRUE)`, where the output type id column for v3 samples effectively contains the compound_idx. The hashes are calculated on the relevant subsets of values of each sample and aggregated/counted at the relevant level for each check, ie:
- check_tbl_spl_n()` count unique output type id values per compound idx. The hash table provides a mapping between output type ids and compound idxs.
- `check_tbl_spl_compound_tid()`: Ensure there is only a single unique hash of the combination of values across compound task id columns of all rows associated with samples for a given compound idx.
- `check_tbl_spl_non_compound_tid()`: Ensure there is only a single unique hash of the combination of values across non-compound task id columns of all rows associated with samples for a modeling task.

These checks are performed separately for each round modeling task item, allowing for differences between compound task id sets between round modeling tasks.

### Still to do:

- [ ] The `spl_hash_tbl()` on which many of the new checks depend can be time consuming with complex configs. Have attempted memoisation but have encountered difficulties in testing so this is still a work in progress but shouldn't change the rest of the functionality.
- [ ] Add more tests, especially varying the compound task id set.

elray1 commented on DESCRIPTION:
```diff
@@ -60,9 +60,9 @@ Suggests:
-    Infectious-Disease-Modeling-Hubs/hubUtils,
-    Infectious-Disease-Modeling-Hubs/hubData,
-    Infectious-Disease-Modeling-Hubs/hubAdmin,
+    Infectious-Disease-Modeling-Hubs/hubUtils@enhancement/v3-utils,
+    Infectious-Disease-Modeling-Hubs/hubData@feature/handle-samples,
+    Infectious-Disease-Modeling-Hubs/hubAdmin@feature/sample-support,
```
noting that we will probably want to change this before going ahead with a merge

annakrystalli replied:
Most definitely.
It's the last thing that needs to be done throughout the packages. It is required atm for tests to run successfully.

LucieContamin reviewed (changes requested):
I have some questions about the `check_tbl_spl_n()` behavior. I also have some minor documentations questions. 

I was also debating if it makes sense to have somewhere in the documentation (function documentation and/or vignette) a warning saying that large files with a lot of samples might take time to validation. However, as we don't have a clear estimation on what is "large" and "take time", I am not sure how helpful it is. 

In my review, I refer to a test that I made where I had issue with the results. 
For the test, I use my own temporary "test" repo at LucieContamin/hub_test and the file causing issue is `"Hubtest-hubtemp_subset/2024-04-28-Hubtest-hubtemp_subset.parquet"`

annakrystalli commented:
> I was also debating if it makes sense to have somewhere in the documentation (function documentation and/or vignette) a warning saying that large files with a lot of samples might take time to validation. However, as we don't have a clear estimation on what is "large" and "take time", I am not sure how helpful it is.

While it would be useful, I also agree that as "large" and "take time" are hard to properly define, not sure just how useful.  There are plans to try and improve the performance of validations though so as part of that work we might get a better sense of what would be useful in the documentation also? I'll draft the performance issue up today and make a note about including more on performance in the docs too.

annakrystalli commented on R/check_tbl_spl_compound_tid.R:
```diff
@@ -0,0 +1,81 @@
+    details <- cli::format_inline(
+      "Sample{?s} {.val {mismatches}} d{?oes/o} not match most prevalent ",
+      "task ID combination for {?its/their} compound idx. ",
+      "See {.var errors} attribute for details."
```
So the reason I've used "prevalent" has to do with the fact that, within a given submission file, when providing samples, folks do not need to include predictions for all optional values of non-compound task id values. They do however need to include predictions for **the same** optional values across samples (i.e. combinations of values in samples for each compound_idx need to be consistent across all samples).  This is harder to validate than it sounds as there's effectively no exact baseline to compare all samples against. If they are all the same, great, validation passes but if not, how do we decide which samples are incorrect? The way I decided to approach it is to check the frequency of different combinations, take the most frequent combination as the baseline and then declare ones that do not match the baseline as incorrect. This is what I mean by "prevalent". 

Say for example in your test hub that horizons > 10 were optional.  Most sample contain predictions for horizons 1:15 but in a couple of samples, predictions for horizons 1:20 are included.  They are both technically correct value combinations but because the majority of samples only give horizons 1:15, and only a couple deviate from that, the ones that deviate from the most prevalent (i.e. common) task id combination in the sample are flagged as erroneous.

Does the explanation make sense. I think the confusion is justified and was also not sure about the word "prevalent" but it was the best I could come up with. Any suggestions on how to get this message across welcome!!

LucieContamin replied:
Don't we have a baseline with the information in the associated json?
For example, if we have 6 tasks ids column: `origin_date`, `scenario_id`, `location`, `target`, `horizon`, `age_group` and with a mimimum of 100 samples expected with `"compound_taskid_set": ["origin_date", "scenario_id", "location", "target"]` (I am going to keep the same example all along), is it not expected to have at the minimum for each sample: all the unique horizon and age group possible value, at the minimum? 

|origin_date|scenario_id|location|target|horizon|age_group|output_type_id| compound_idx|
|---|---|---|---|---|---|---| ---|
|2024-04-28| A | US| inc hosp | 1 |65-130|s1|0|
|2024-04-28| A | US| inc hosp | 1 |65-130|s2|0|
...
|2024-04-28| A | US| inc hosp | 1 |65-130|s100|0|
|2024-04-28| A | US| inc hosp | 2 |65-130|s1|0|
|2024-04-28| A | US| inc hosp | 2 |65-130|s2|0|
...
|2024-04-28| A | US| inc hosp | 2 |65-130|s100|0|
...
...
|2024-04-28| A | US| inc hosp | 1 |0-64|s1|0|
|2024-04-28| A | US| inc hosp | 1 |0-64|s2|0|
...
|2024-04-28| A | US| inc hosp | 1 |0-64|s100|0|
...

But a model can submit a file with another level of "compound_task_id_set" (for example: `["origin_date", "location", "target"]`) and it should still be OK because the minimal set will be respected. Here, each sample will have all the unique scenario, horizon and age group possible value.

LucieContamin replied:
I mean we can check that the minimum is respected and that each sample follow the same compound task id set

annakrystalli commented on tests/testthat/_snaps/check_tbl_spl_compound_tid.md:
```diff
@@ -0,0 +1,40 @@
+    Output
+      <warning/check_failure>
+      Warning:
+      Task ID combinations across compound idx samples not consistent.  Sample "1" does not match most prevalent task ID combination for its compound idx. See `errors` attribute for details.
```
Totally agree! I don't think it will make it into this release but I've actually got a related issue open on this: https://github.com/Infectious-Disease-Modeling-Hubs/hubValidations/issues/76

LucieContamin replied:
Perfect, sorry I miss the already open issue

annakrystalli commented on R/check_tbl_spl_n.R:
```diff
@@ -0,0 +1,124 @@
+  hash_tbl <- spl_hash_tbl(tbl, round_id, config_tasks)
+  n_ranges <- get_round_spl_n_ranges(config_tasks, round_id)
+
+  n_tbl <- dplyr::group_by(hash_tbl, .data$compound_idx) %>%
+    dplyr::summarise(
+      n = dplyr::n_distinct(.data$output_type_id),
+      mt_id = unique(.data$mt_id)
+    ) %>%
+    dplyr::left_join(n_ranges, by = "mt_id") %>%
+    dplyr::mutate(
+      less = .data$n < .data$n_min,
+      more = .data$n > .data$n_max,
+      out_range = .data$less | .data$more
+    ) %>%
+    dplyr::filter(.data$out_range)
```
Right. Compound_idx is what defines a sample and we need to count samples across compound_idx to ensure they are within the correct range. See for eaxample how samples (output_type_ids) are marked up under different coumpound task id sets in the docs: https://hubverse.io/en/latest/user-guide/sample-output-type.html#four-submissions-differing-by-compound-modeling-task

annakrystalli replied:
I've also opened the following issue. You can currently get this infomration from the hubData::expand_model_out_val_grid(include_sample_ids = TRUE)` but I think a function that creates a clear table illustrating compound idx structure of a given round/modeling task would be really useful https://github.com/Infectious-Disease-Modeling-Hubs/hubData/issues/40

LucieContamin replied:
The part that I don't understand is why is it link to the output type id information? For me it's two different information. We are checking the output type id column in the other 2 functions with the compound ID checking, so if you have an error of ID, you will have an error but it does not mean you are missing samples. 
For example if someone provided the correct number of samples but made an error in the output type id column, than it should not return an error here but only about the output type id numbering.

annakrystalli replied:
In the validation instriuctions it states:
> The number of rows for each combination of individual modeling task should fall between min_samples_per_task and max_samples_per_task (inclusive).

Compound_idx defines the individual modeling task of the combination unique compound task id values, so samples need to match a modeling task first to then be counted against the required number for that compound_idx.

I you have provided the wrong output type ID, that should cause an error here too because you have not properly specified the samples, so it would be wrong in view to count them as valid samples.

From the validation spec sheet, this the test I've tried to encode here:
> - Example: task ids are location, origin_date, horizon; compound units defined by combinations of location and origin_date; expect the number of samples for each such compound unit to be between specified min and max.
> - Logic: Within each group defined by a unique combination of values for the compound_taskid_set, the number of unique values for output_type_id should be between min_samples_per_task and max_samples_per_task

LucieContamin replied:
I am sorry I miss and/or misunderstood some of the part in the instructions for the validation tests, for me the issue here in the "combination of individual modeling task" is are we talking about the minimum compound set the hub is asking or the compound set chosen by the team who made the submission?

annakrystalli replied:
Don't worry! It is confusing...I'm confused  !

My interpretation is that combination of individual modeling tasks is defined by the `compund_taskid_set` in the config.  Each unique combination of values in the variables in the `compund_taskid_set` corresponds to a single `compound_idx`. And number of samples (i.e. unique `output_type_ids`) is counted separately for each `compound_idx`.

I note as well that some of this confusion is probably coming from the fact that in the description we do mention coarser samples passing validation (as you noted below) but I haven't implemented that as it wasn't in the validation instructions.

See below for my suggestion of how to proceed with this for now.

LucieContamin replied:
Interestingly, I tested the validation with a coarser sample settings and it passes validation without any issue (JHU subset file) if the same compound id set is used everywhere. 
My hypothesis (but I am very not sure), is that it's because in the `get_mt_spl_hash_tb()` and in particular in the 
```r
split(tbl, f = tbl$output_type_id) %>%
    purrr::map(
      function(.x, compound_taskids, non_compound_taskids) {
        tibble::tibble(
          compound_idx = names(sort(table(.x$compound_idx), decreasing = TRUE))[1L],
          output_type_id = unique(.x$output_type_id),
          hash_comp_tid = rlang::hash(unique(.x[, compound_taskids])),
          hash_non_comp_tid = rlang::hash(.x[, non_compound_taskids]),
          hash_spl_id = rlang::hash(.x)
        )
      },
      non_compound_taskids = non_compound_taskids,
      compound_taskids = compound_taskids
    ) %>%
    purrr::list_rbind()
```
If all the samples share the same first compound id then it passes without issue.

LucieContamin replied:
> I you have provided the wrong output type ID, that should cause an error here too because you have not properly specified the samples, so it would be wrong in my view to count them as valid samples against the expected number.

Also, I understand the logic behind the error but I also wonder if it's confusing for the user. 
For example, I still don't understand the output I got from my test. I have the expected number of sample, but hey are indeed misidentified. However, as they fit the minimal requirement of 100 samples per the hub compound id set test, I have no idea how to fix it. As we say before the error message might also need more information because for example I got and I don't know what the compound idx 10 means here as it's associated with different output type id:
```
Required samples per compound idx task not present.
  File contains less ("69") than the minimum required number of samples per task ("100") for
  compound idx "10" [...]
```

annakrystalli replied:
If you dig into the `errors` object, it gives you a lot more detail including what the composition of each `compound_idx` and how many samples of that idx were successfully counted, eg. here's the first element of the output:
```r
$compound_idx
[1] "10"

$n
[1] 69

$min_samples_per_task
[1] 100

$min_samples_per_task
[1] 300

$compound_idx_tbl
# A tibble: 39,312 × 6
   origin_date scenario_id  location target   horizon age_group
   <chr>       <chr>        <chr>    <chr>    <chr>   <chr>    
 1 2024-04-28  A-2024-03-01 51       inc hosp 1       0-64     
 2 2024-04-28  A-2024-03-01 51       inc hosp 2       0-64     
 3 2024-04-28  A-2024-03-01 51       inc hosp 3       0-64     
 4 2024-04-28  A-2024-03-01 51       inc hosp 4       0-64     
 5 2024-04-28  A-2024-03-01 51       inc hosp 5       0-64     
 6 2024-04-28  A-2024-03-01 51       inc hosp 6       0-64     
 7 2024-04-28  A-2024-03-01 51       inc hosp 7       0-64     
 8 2024-04-28  A-2024-03-01 51       inc hosp 8       0-64     
 9 2024-04-28  A-2024-03-01 51       inc hosp 9       0-64     
10 2024-04-28  A-2024-03-01 51       inc hosp 10      0-64     
# ℹ 39,302 more rows
# ℹ Use `print(n = ...)` to see more rows
```

There's only so much detail I can add to the message (in fact looking at it it's too long atm!). Also after I've implemented https://github.com/Infectious-Disease-Modeling-Hubs/hubData/issues/40, users will have another eay to inspect expected hub compound idxs

LucieContamin replied:
So sorry I was not clear, I dig into the errors message but still cannot find how to replicate the n=69, I know the solution here is to update the sample ID in the output type id but it was just to illustrate my comment on how I find it confusing.

annakrystalli commented on R/check_tbl_spl_n.R:
```diff
@@ -0,0 +1,124 @@
+        setdiff(names(tbl), c("output_type_id", "value", "output_type"))
+      ]
+      row <- n_tbl[n_tbl$compound_idx == .x, ] %>% as.vector()
+      list(
+        compound_idx = .x,
+        n = row$n,
+        min_samples_per_task = row$n_min,
+        min_samples_per_task = row$n_max,
+        compound_idx_tbl = compound_idx_tbl
```
To ensure number of samples are correct, we can't just count the number of distinct values across the compound task id set. We need to ensure that they have been correctly allocated to the appropriate sample through the value in output type id. And each sample (i.e. unique `output_type_id` value) should have only a single value in each of the compound task id set columns:

In the file you have been testing, the samples do not appear to be correctly specified, for example, sample no 10 (with output_type_id == 10) contains more than one unique value for `scenario_id` when it should contain only one:
```r
tbl |> filter(output_type_id == 10) |> select( origin_date, scenario_id,  location, target) %>% distinct()
# A tibble: 276 × 4
   origin_date scenario_id  location target  
   <chr>       <chr>        <chr>    <chr>   
 1 2024-04-28  A-2024-03-01 51       inc hosp
 2 2024-04-28  B-2024-03-01 51       inc hosp
 3 2024-04-28  C-2024-03-01 51       inc hosp
 4 2024-04-28  D-2024-03-01 51       inc hosp
 5 2024-04-28  E-2024-03-01 51       inc hosp
 6 2024-04-28  F-2024-03-01 51       inc hosp
 7 2024-04-28  A-2024-03-01 05       inc hosp
 8 2024-04-28  B-2024-03-01 05       inc hosp
 9 2024-04-28  C-2024-03-01 05       inc hosp
10 2024-04-28  D-2024-03-01 05       inc hosp
# ℹ 266 more rows
# ℹ Use `print(n = ...)` to see more rows
```

This is throwing off the `sample_n` validation and should have been picked by `check_tbl_spl_compound_tid()` but sadly this is causing an error.

You can investigate the error by looking at the `error` object (as discussed this could be made clearer). Also having a look at the `sample` subset of the`expand_model_out_val_grid(...,  include_sample_ids = TRUE)` output might be able to help in the meantime to see what samples for different compound idxs should look like (until https://github.com/Infectious-Disease-Modeling-Hubs/hubData/issues/40 is implemented.)

LucieContamin replied:
>  And each sample (i.e. unique output_type_id value) should have only a single value in each of the compound task id set columns:

I thought we were allowing different level of compound id set. For example, in the documentation (https://hubverse.io/en/latest/user-guide/sample-output-type.html#configuration-of-output-type-id) the compound set C is accepted with all the other compound set. 

So here if I have a team that submit a file with another level of `"compound_task_id_set"` (for example: `"origin_date", "location", "target"]`), it should be accepted and different scenario id value should be expected. I tested it with a file where the compound id set is: `["origin_date"]` and it worked (`JHU_UNC-flepiMoP_subset` in my example repo).

annakrystalli replied:
Hmmm indeed. However this wasn't explicit in the[ instructions for the validation tests](https://docs.google.com/document/d/1vJT8Uw8fQgoY5sajcokZSfnF7M89eiiiwlRXVkCuj2k/edit#heading=h.qli9sktijlj4). It also makes validating a lot more complicated and would need thought as to how to perform such a loose check. 

At the moment, as the tests are set up this is not allowed and I don;t think it will be simple enough to introduce before the planned release so perhaps opening an issue to include tis and think through how it would even work would be the best way to go forward atm.

annakrystalli replied:
Also origin date would pass because there is normally only a single value per submission file.

LucieContamin replied:
If we want a quick test that does not take into account the `output_type_id`, you can use the code I use here to calculate the number of samples per each task id group (without taking into account the `"compound_task_id_set"`) but we loose the link with the compound task id. However it will allow tag that the error is just on the ID of the sample not on the total number of samples in the submission file. If we want to provide that information

annakrystalli replied:
Honestly, I still don't feel that's right. But I think the problem here is that when files have not passed the validation checks for correctly defined samples, trying to count the samples is pointless. I therefore think the  correct way to proceed would be to set up tests so that they are passed sequentially.

annakrystalli commented:
Firstly thanks so much for your review and thorough testing @LucieContamin ! It's been really useful to work through. In response I've made a number of changes to the functionality / docs:

- Firstly, to make things more streamlined, I've changed the sequence of execution of sample checks and set the check for the compound task id and non-compound task id to return errors and cause validation to return early. That way samples are only counted once we know we have well formed samples (I know it's different to what you do in scenario hub but it makes more sense to me atm, happy to revisit and get more opinions in next round of work on samples though!).
- Next I've reworked the check messages, the names of the objects returned if validation fails and the information returned as well as part of each `errors`. Hopefully, the information returned is much more useful and intuitive and the information in the docs is now enough to help explain what each check failure means and direct a tema to fixing it.
- I've also opened this issue to add functionality to validate coarser compound task id sets which atm `hubValidations` does not support (https://github.com/Infectious-Disease-Modeling-Hubs/hubValidations/issues/88). I have added questions to that issue to help me understand the functionality better, your input would be greatly appreciated!

Let me know if these resolve your issues for the time being and feel free to open more issues if you think there's more that needs to be addressed.

LucieContamin commented on R/check_tbl_spl_non_compound_tid.R:
```diff
@@ -0,0 +1,102 @@
+      unlist(use.names = FALSE)
+
+    details <- cli::format_inline(
+      "Sample{?s} {.val {output_type_ids}} d{?oes/o} not match most prevalent ",
+      "non compound task ID combination for {?its/their} modeling task. ",
+      "See {.var errors} attribute for details."
```
As the compound ID error message was updated, should we also update that one so that they "match"? Maybe something like: 
```
Sample{?s} {.val {output_type_ids}} d{?oes/o} not match ",
      "the expected non compound task ID combinations.",
      "See {.var errors} attribute for details."
 ```

annakrystalli replied:
Here it is comparing non compound task id combinations (which can include optional values - i.e. in this test there is no baseline to check against but a number of potential combinations across all rows of a sample which all samples need to share) to the most prevalent combination and identifying ones that don't match. So this is actually an accurate description of what the check is doing.

LucieContamin replied:
I thought about it because the error message changed for the other function (which makes sense), but also the documentation of this function changed to `"Check model output data tbl samples contain single unique combination of non-compound task ID values across all samples"` so I thought it was making sense to update the error message to "match" non only with the other function style but also the title of the documentation here. I see now that `expected` was not the best description. Also if I am the only one having trouble with `prevalent` here, feel free to ignore. I was confused by it before, but after more explanation, `prevalent` makes sense to me now, so my only hesitation was that I was not sure if it would make sense for the users.

annakrystalli replied:
I don't feel error messages need to repeat the same information throughout verbatim. The samples do need to be consistent but it is important to state how we determine that consistency and how we identify which are the deviations from that consistency. Because of that it is important to mention that we compare to the most prevalent combination. Having said that, because it might just be the word "prevalent" that is the problem, I've changed it to "frequent" throughout which hopefully is more widely used and understood?

LucieContamin replied:
Perfect, thank you very much.

LucieContamin commented on R/check_tbl_spl_n.R:
```diff
@@ -0,0 +1,140 @@
+    switch(type,
+      less = paste0(
+        "File contains less than the minimum required number of samples per task ",
+        "for compound idx{?s} {.val {compound_idx}}"
```
As `compound_idx` was change to `output type ids` is the other error messages, will it make sense to do it here too? (if it's not too difficult)

annakrystalli replied:
Here the check is counting samples per compound_idx so the message is again accurate.

LucieContamin replied:
Ah yes sorry I was not clear, the check is indeed counting samples per compound_idx so that makes sense. However, the compound_id is not something the user has access to easily. So, I was wondering if it makes sense to simplify which compound id we are talking about (maybe with which columns value it represents, for example: Scenario A, location US, etc.).
Also I realize now that adding the output type id information here might not be the most helpful. Sorry about that

annakrystalli replied:
Users can access compound_idx using `hubData::expand_model_out_val_grid(..., include_sample_ids = TRUE)` but they also will be when the function that creates a clear table illustrating compound idx structure of a given round/modeling task is available: https://github.com/hubverse-org/hubData/issues/40
I am also going to suggest we move both those functions to `hubValidations` so they are available to users.

More importantly the details of each `compound_idx` structure is available in the `errors` object so users can immediately see what is expected for a sample to count towards a particular `compound_idx`. 

So I would prefer to leave the mention to compound_idx for now as I would only have to re-introduce when https://github.com/hubverse-org/hubData/issues/40 is complete (although when it is complete I will mention it in the documentation)

LucieContamin replied:
OK sounds good to me, thanks for the additional information!

LucieContamin reviewed (commented):
Thank you for all the updates! To respond to your three points updates:
1. (streamline) I think it's a very good idea. It makes the validation faster
2. (error message) I ran some tests on the new version and the error I got is easier to understand, thanks again!
3. (coarser compound id) The new version of the validations does not support correctly ID sample for coarser compound task id group anymore (it was in the first reviewed version). I understand  why but wonder if it's going to be something we need to add relatively quickly. I will comment in the open issue.

I also add some minor comment on the output error messages.
